mesos-client/src/main/scala/com/mesosphere/mesos/client/CredentialsProvider.scala

```diff
@@ -0,0 +1,93 @@
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.{BasicHttpCredentials, GenericHttpCredentials, HttpCredentials}
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.ActorMaterializer
+import com.typesafe.scalalogging.StrictLogging
+import de.heikoseeberger.akkahttpplayjson.PlayJsonSupport
+import pdi.jwt.JwtJson
+import pdi.jwt.JwtAlgorithm.RS256
+import play.api.libs.json._
+import play.api.libs.json.Reads._
+
+import scala.async.Async.{async, await}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+
+/**
+  * Base credential provider for [[MesosClient]].
+  */
+trait CredentialsProvider {
+
+  /** @return the next [[HttpCredentials]]. This might be a noop, eg for [[BasicAuthenticationProvider]]. */
+  def nextToken(): Future[HttpCredentials]
+}
+
+/**
+  * Provides DC/OS session tokens as JWT for service accounts.
+  *
+  * @param uid The user id.
+  * @param privateKey The JWT secret.
+  * @param root The DC/OS root URL.
+  * @param system The [[ActorSystem]] used for HTTP requests.
+  * @param materializer The [[ActorMaterializer]] use to unmarshal HTTP responses.
+  * @param context The execution context for transforming the future HTTP response.
+  * @see See [[https://docs.mesosphere.com/1.11/security/ent/iam-api/#obtaining-an-authentication-token]] for details on
+  *      the protocol.
+  */
+case class DcosServiceAccountProvider(uid: String, privateKey: String, root: URL)(
+    implicit system: ActorSystem,
+    materializer: ActorMaterializer,
+    context: ExecutionContext)
+    extends CredentialsProvider
+    with PlayJsonSupport
+    with StrictLogging {
+
+  private def expireIn(duration: Duration): Long = Instant.now.getEpochSecond + duration.toSeconds
+
+  /**
+    * Construct a claim to authenticate and retrieve a session token. The timeout is for this token ''not'' the session
+    * token that we are requesting. That means if ou login request takes more than twenty seconds the authentication will
+    * fail.
+    *
+    * @return a claim for the user that expires in twenty seconds.
+    */
+  private def claim: JsObject = Json.obj("uid" -> uid, "exp" -> expireIn(20.seconds))
+
+  case class SessionToken(token: String)
+  implicit val sessionRead: Reads[SessionToken] = (JsPath \ "token").read[String].map(SessionToken)
+
+  /** @return a new request for a session token. */
+  def acsTokenRequest: HttpRequest = {
+
+    // Tokenize the claim.
+    val token = JwtJson.encode(claim, privateKey, RS256)
+    val data: String = Json.stringify(Json.obj("uid" -> uid, "token" -> token))
+
+    HttpRequest(
+      method = HttpMethods.POST,
+      uri = Uri(s"$root/acs/api/v1/auth/login"),
+      entity = HttpEntity(ContentTypes.`application/json`, data)
+    )
+  }
+
+  override def nextToken(): Future[HttpCredentials] = async {
+    logger.debug(s"Fetching next token from $root")
```

Comment:
Since this happens every ~20min I'd increase logging level:
```suggestion
    logger.info(s"Fetching next token from $root")
```